sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def inputTypes: Seq[DataType] = {
+    // TODO: properly handle String Types with differing collations
+    val leftIsString = left.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    val rightIsString = right.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
+
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
+    // TODO: ensure that lCollationID and rCollationID are equal (compatible)
+    val collationID = if (lCollationID > rCollationID) lCollationID else rCollationID
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    collator match {
+      // UTF8String collation
+      case null =>
+        collationID match {
+          // UCS_BASIC
+          case 0 =>
+            l.contains(r)
+          // UCS_BASIC_LCASE
+          case 1 =>
+            // scalastyle:off caselocale
+            l.toLowerCase.contains(r.toLowerCase)
```

Comment:
Note: This is for UCS_BASIC_LCASE - the only collation type, apart from UCS_BASIC, that doesn't have a "collator".

Reply:
Yeah, I think that `collator` should be private and shouldn't be exposed outside of Collation. Let's add new function if we need anything special (e.g. `normalize` or even `contains`).

Reply:
UCS_BASIC_LCASE seems like a special case. "normalize" in the context of UCS_BASIC_LCASE seems to always be "toLowerCase", and these binary built-in string functions for UCS_BASIC_LCASE generally seem to amount to `l.toLowerCase.<str_fun>(r.toLowerCase)` (instead of `l.<str_fun>(r)` which is the case for UCS_BASIC).

On the other hand, adding `contains` in Collation seems like a good way to separate things out!

Reply:
Update: added `contains`, `startswith`, `endswith` directly to Collation in new [PR](https://github.com/apache/spark/pull/45216/files)